### DISCUSSION TITLE
OSError: Could not autodiscover CAN Channel in Linux / Raspberry Pi

### DISCUSSION THREAD

**Discussion by adbmx** [tinymovr/Tinymovr#175] [Q&A - Tinymovr] [answered]
Hi 

When trying to launch tinymovr, the launch process aborts with the error message: `OSError: Could not autodiscover CAN channel`

I installed tinymovr as described in "[From Zero to Spin](https://tinymovr.readthedocs.io/en/latest/quickstart.html)" on the Raspberry Pi and with conda on the host machine.
I flashed a new firmware on the CANine adpater as [described in readthedocs](https://canine.readthedocs.io/en/latest/canine.html#flashing-firmware). The adapter now appears with lsusb as `Bus 001 Device 002: ID 0483:c1b0 STMicroelectronics CANine CDC`, so I guess the flashing did work.

I installed tinymovr on 3 different Linux distributions:
- Manjaro 21.2.4
- Mint 20.2
- Raspbian 10

The hardware setupt is as follows:
- Raspberry Pi (or host computer)
- CANine adpater
- 2 tinymovr boards

I tried to connect to a single and to both tinymovr boards at the same time. Cables are all ok and plugs are in the right spot. Switches on the tinymovr boards are as they should be. I tried the setup with a single and two terminal resistors, there is no influence. I am not sure if the tinymvor board gets detected by the CANine adapter. I am new to can bus and to tinymovr. This was actually the first time I tried to launch the system.

Here is the whole error message:

```
Traceback (most recent call last):
  File "/opt/anaconda/envs/tinymovr/bin/tinymovr", line 8, in <module>
    sys.exit(spawn_shell())
  File "/opt/anaconda/envs/tinymovr/lib/python3.7/site-packages/tinymovr/shell.py", line 67, in spawn_shell
    channel = guess_channel(bustype_hint=bustype)
  File "/opt/anaconda/envs/tinymovr/lib/python3.7/site-packages/tinymovr/iface/can_bus/can_bus.py", line 95, in guess_channel
    raise IOError("Could not autodiscover CAN channel")
OSError: Could not autodiscover CAN channel
```

Any ideas? What am I overseeing?

**Comment by yconst** [comment]
If I understand correctly, you are using the CANine-specific firmware at https://github.com/tinymovr/CANine ? 

This firmware requires a fork of Python-CAN to be installed, otherwise it is not detected. In addition, <del>development of this firmware has stagnated due to focus on the main Tinymovr fw/sw development</del> (the [CANine firmware has matured and is available here](https://github.com/tinymovr/CANine)). For the Raspberry Pi, I would highly suggest using Candlelight, which is supported out of the box as described here: https://tinymovr.readthedocs.io/en/latest/studio/usage.html?highlight=candlelight#socketcan-linux

**Comment by adbmx** [comment, marked as answer]
Hi Yannis

Thank you for your quick reply. I could solve the problem.
Yes, I flashed the CANine adapter with the CANine-specific firmware. With the CANine-specific firmware no device node in `/dev` was found (missing driver).

After [updating the CANine adapter to a new firmware](https://canable.io/updater/) (slcan), the adapter was recognized on the host computer, but not on the raspberry pi. 
With the new firmware installed I was able to launch tinymovr studio on the host machine with the [compatibility message](https://tinymovr.readthedocs.io/en/latest/studio/usage.html?highlight=Compatibility#compatibility): 
`Node fw version incompatible (found: 0.8.11, required: 0.8.12)`.
Despite the new firmware on the CANine adapter, the error message `OSError: Could not autodiscover CAN channel` still showed up on the raspberry pi when launching tinymovr studio.

After downgrading tinymovr to version 0.3.11 with `pip install tinymovr==0.3.11` on both, the host machine and the raspberry pi, I was able to launch tinymovr studio on both systems and everything works fine now.

With the candlelight firmware I still get the error message `OSError: Could not autodiscover CAN channel` on both systems.
With the slcan firmware the adapter appears with lsusb as follows: `Bus 001 Device 013: ID ad50:60c4 Protofusion Labs CANable 1205aa6 https://github.com/normaldotcom/cantact-fw`
At the moment I need to start the tinymovr studio with the sudo command on the host machine, but I think with the right udev rule this won't be necessary in the future.

I think the reason why the raspberry pi didn't show the compatibility message with tinymovr studio 0.3.12 and slcan firmware is because the `desc_lower` variable in `guess_channel()` function returns only the device node name `ttyacm0` an therefore can't find a match in `can_devices`.

Best regards
